flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -708,22 +689,33 @@ public void receiveDeclineMessage(DeclineCheckpoint message, String taskManagerL
 				return;
 			}
 
-			checkpoint = pendingCheckpoints.remove(checkpointId);
+			checkpoint = pendingCheckpoints.get(checkpointId);
 
-			if (checkpoint != null && !checkpoint.isDiscarded()) {
+			if (checkpoint != null) {
+				Preconditions.checkState(
+					!checkpoint.isDiscarded(),
+					"Received message for discarded but non-removed checkpoint " + checkpointId);
 				LOG.info("Decline checkpoint {} by task {} of job {} at {}.",
 					checkpointId,
 					message.getTaskExecutionId(),
 					job,
 					taskManagerLocationInfo);
-				discardCheckpoint(checkpoint, message.getReason(), message.getTaskExecutionId());
-			}
-			else if (checkpoint != null) {
-				// this should not happen
-				throw new IllegalStateException(
-						"Received message for discarded but non-removed checkpoint " + checkpointId);
-			}
-			else if (LOG.isDebugEnabled()) {
+				final CheckpointException checkpointException;
+				if (message.getReason() == null) {
+					checkpointException =
+						new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED);
+				} else if (message.getReason() instanceof CheckpointException) {
+					checkpointException = (CheckpointException) message.getReason();
+				} else {
+					checkpointException =
+						new CheckpointException(
+							CheckpointFailureReason.JOB_FAILURE, message.getReason());
+				}
```

Comment:
extract to a helper method?

Reply:
OK

Reply:
Has this happened? I can not find this fixed in the the last `[hotfix] Address review comments.` nor it's fixed in this commit.

Btw, having one commit with all of the addressed comments, where comments were referring to multiple commits is not the best solution. Since you would have to fixup the correct commits before merging, it would be better to have couple of smaller fixup commits (one fixup commit per original commit). (Also FYI, I would be fine with you already squashing the fixup commits with the original commits)

Reply:
Ah, I forgot to leave an explanation about this. I didn't squash the original commits to make it easy to continue reviewing.

Actually I have extracted the "else if ... else" to a helper method named `getCheckpointException`. But it does happen in fixup commit but in the original commit "[FLINK-13905][checkpointing] Separate checkpoint triggering into several asynchronous stages.". I thought it's enough as a helper method.

And thanks for the reminding to keep per fixup as a separate commit. I'll try to separate the last big fixup commit :)